doc/amazon_sagemaker_model_building_pipeline.rst

```diff
@@ -0,0 +1,872 @@
+
+Parameterization
+================
+
+Being able to use interfaces from other existing SageMaker python SDK modules, along with the Pipeline module offers us great convenience and flexibility to construct a pipeline. Furthermore, parameterization is crucial for pipeline composition, it allows us to assign values to properties that are not available until runtime. However, there are limitations.
+
+Pipeline parameterization includes pipeline parameter (e.g. :code:`ParameterString`, :code:`ParameterInteger`, and etc.), property reference, functions (:code:`Join` and :code:`JsonGet`), and pipeline execution variables. Pipeline Parameterization might not be supported with 100% compatibility when used together with other SageMaker python SDK modules.
+
+Many args for class constructors or methods from other modules can be parameterized, but not all of them. For instance, when construct a Processor, :code:`instance_type` or :code:`instance_count` can be parameterized; when calling :code:`processor.run`, inputs or outputs can also be parameterized.
+
+.. code-block:: python
+
+    instance_count = ParameterInteger(name="InstanceCount", default_value=2)
+    instance_type = ParameterString(name="InstanceType", default_value="ml.m5.xlarge")
+    process_s3_input_url = ParameterString(name="ProcessingInputUrl")
+
+    processor = Processor(
+       instance_type=instance_count,
+       instance_count=instance_type,
+       ...
+    )
+    processor.run(inputs=ProcessingInput(source=process_s3_input_url), ...)
+
+Nonetheless, for example, one cannot parameterize :code:`git_config` when calling :code:`processor.run`. This is due to the source code needs to be downloaded, packaged, and uploaded to s3 at compile time, yet the parameterization can only be evaluated at runtime.
+
+Another common example is applying python built-in operation on a pipeline parameter.  For instance, one might try to concatenate the pipeline variables with Python primitives like this:
+
+.. code-block:: python
+
+    my_string = "My{}Step".format(ParameterString(name="StepMiddleName", default_value=""))
+
+This won’t work as the parameter :code:`StepMiddleName` can be known only at runtime. Instead, the same concatenation can be achieved by :code:`Join`
+
+.. code-block:: python
+
+    my_string = Join(on="", values=[
+        "My",
+        ParameterString(name="StepMiddleName", default_value=""),
+        "Step"]
+    )
+
+In summary, pipeline parameter can only be evaluated at runtime, if its actual value needs to be known at compile time in order to either conduct business logic, or apply some operations against, it will lead to an exception.
+
+====================================
+Pipeline Composition Alternative
+====================================
+
+Python SDK provides us with a great abstraction for pipeline composition. Under the hood, it produces a pipeline definition JSON. If you want to author the pipeline definition by hand, you might follow this `JSON Schema`_
```

Comment:
trailing underscore?

Reply:
this is the link reference, you do `X`_, and then in another line, : .X: https://google.com